server/test/unit/plugins/file/test_distributor.py

```diff
@@ -345,3 +345,83 @@ def test__symlink_units_EINVAL_os_error(self, mock_readlink):
         self.assertTrue(os.path.islink(original_link))
         created_link = readlink(original_link)
         self.assertNotEqual(old_target, created_link)
+
+    @patch('os.path.join')
+    @patch('os.path.isabs')
+    @patch('os.path.normpath')
+    def test__target_symlink_path_OK_path(self, mock_normpath, mock_isabs, mock_join):
+        """
+        Check an OK target path passes the symlink path validation
+        """
+        distributor = self.create_distributor_with_mocked_api_calls()
+        build_dir = '/foo/bar'
+        target_path = 'baz'
+        # set up mock return values according to an OK path
+        norm_mock = mock_normpath.return_value
```

Comment:
I like the way you patch the os.path functions while retaining their original implementation below (using `side_effect`). Couldn't we use it here as well? i.e.:

```python
@patch('os.path.join', side_effect=os.path.join)
@patch('os.path.isabs', side_effect=os.path.isabs)
@patch('os.path.normpath', side_effect=os.path.normpath)
def test__target_symlink_path_OK_path(self, mock_normpath, mock_isabs, mock_join):
...
```

Then you would not need to mock all these values and could still used the `assert_called_...`functions like you do in the other tests.

Reply:
I stole the idea from the [test code above](https://github.com/pulp/pulp/blob/ae11e67274db41792054ffb8682c7057ab97b4bd/server/test/unit/plugins/file/test_distributor.py#L268), I like it too :) My goal here was to assert that all the "gears wheeled" as expected, esp. the `normpath()` return value equality with the expected `..` parameter,  thru the `norm_mock.__eq__` magic method. The `side_effect=os.path.normpath` would make the `mock_normpath` return a string instead of a mock, I wouldn't be able to check the gears on this level anymore.

Reply:
Ok, I see where you are coming from.